domain/application/resource/resourcestore.go

```diff
@@ -5,36 +5,80 @@ package resource
 
 import (
 	"context"
+	"fmt"
 	"io"
-	"sync"
+	"path"
 
+	"github.com/juju/juju/core/application"
 	"github.com/juju/juju/core/objectstore"
-	"github.com/juju/juju/domain/application/errors"
 	"github.com/juju/juju/internal/charm/resource"
 )
 
 // ResourceStore provides a list of methods necessary for interacting with
 // a store for the resource.
 type ResourceStore interface {
-	// Get returns an io.ReadCloser for data at path, namespaced to the
-	// model.
-	Get(context.Context, string) (io.ReadCloser, int64, error)
+	// Get returns an io.ReadCloser for a resource in the resource store.
+	Get(ctx context.Context, applicationUUID application.ID, name string) (r io.ReadCloser, size int64, err error)
+
+	// Put stores data from io.Reader in the resource store at the
+	// path specified in the resource.
+	Put(
+		ctx context.Context,
+		r io.Reader,
+		applicationUUID application.ID,
+		name string,
+		size int64,
+		fingerprint resource.Fingerprint,
+	) error
+
+	// Remove removes a resource from storage.
+	Remove(ctx context.Context, applicationUUID application.ID, name string) error
 }
 
 // FileResourceStore is a ResourceStore for file resource types.
 type FileResourceStore struct {
 	objectStore objectstore.ObjectStore
 }
 
-func (f FileResourceStore) Get(ctx context.Context, path string) (io.ReadCloser, int64, error) {
+// Get the specified resource from the object store.
+func (f FileResourceStore) Get(
+	ctx context.Context,
+	applicationUUID application.ID,
+	name string,
+) (io.ReadCloser, int64, error) {
+	path := f.storagePath(applicationUUID, name)
 	return f.objectStore.Get(ctx, path)
 }
 
+// Put the given resource in the object store.
+func (f FileResourceStore) Put(
+	ctx context.Context,
+	r io.Reader,
+	applicationUUID application.ID,
+	name string,
+	size int64,
+	fingerprint resource.Fingerprint,
+) error {
+	path := f.storagePath(applicationUUID, name)
+	return f.objectStore.PutAndCheckHash(ctx, path, r, size, fingerprint.String())
+}
+
+// Remove the specified resource from the object store.
+func (f FileResourceStore) Remove(ctx context.Context, applicationUUID application.ID, name string) error {
+	path := f.storagePath(applicationUUID, name)
+	return f.objectStore.Remove(ctx, path)
+}
+
+// storagePath returns the path at which the resource will be stored in the
+// object store. It is stored under the application UUID.
+func (f FileResourceStore) storagePath(applicationUUID application.ID, name string) string {
+	return path.Join("application-"+applicationUUID.String(), "resources", name)
```

Comment:
Imo we should just use the resource UUID for the storage path (and if its soon going to be possible to `Get` from the object store with the object store metadata UUID then we won't even need to save the storage path, we won't need it to access the objects).

In [3.6 they used the storage path](https://github.com/juju/juju/blob/3.6/state/resources.go#L989) as I have copied here. I guess without referential integrity, when a new revision or uploaded version of the resource came along, they would just overwrite the old version in the object store.

Reply:
It'll be a while before the object store Get interface will take a UUID, rather than a path. It's been said that will help with RI and not the rest.

Reply:
Yea, I asked Simon about this yesterday and he said the same, I've just got the UUID there to fetch the resource now.